### ISSUE TITLE
KeyError with Memory widget's MemPercent.

### ISSUE THREAD

**Issue by cjwyett** [qtile/qtile#2012]
When trying to call `widget.Memory(format="{MemPercent}")` I get:

```
Traceback (most recent call last):
  File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.9/site-packages/libqtile/core/manager.py", line 913, in f
    func(*args)
  File "/usr/lib/python3.9/site-packages/libqtile/widget/base.py", line 457, in timer_setup
    update_interval = self.tick()
  File "/usr/lib/python3.9/site-packages/libqtile/widget/memory.py", line 59, in tick
    self.update(self.poll())
  File "/usr/lib/python3.9/site-packages/libqtile/widget/memory.py", line 76, in poll
    return self.format.format(**val)
KeyError: 'MemPercent'
```
Doesn't happen with the other Memory format options.

**Comment by tych0** [comment]
On Wed, Dec 16, 2020 at 04:56:01AM -0800, Cyrus Wyett wrote:
> When trying to call `widget.Memory(format="{MemPercent}")` I get:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
>     self._context.run(self._callback, *self._args)
>   File "/usr/lib/python3.9/site-packages/libqtile/core/manager.py", line 913, in f
>     func(*args)
>   File "/usr/lib/python3.9/site-packages/libqtile/widget/base.py", line 457, in timer_setup
>     update_interval = self.tick()
>   File "/usr/lib/python3.9/site-packages/libqtile/widget/memory.py", line 59, in tick
>     self.update(self.poll())
>   File "/usr/lib/python3.9/site-packages/libqtile/widget/memory.py", line 76, in poll
>     return self.format.format(**val)
> KeyError: 'MemPercent'
> ```
> Doesn't happen with the other Memory format options.

That is in an unreleased version of qtile, you'll need to use git
master for it.

**Comment by tchar** [comment]
> When trying to call `widget.Memory(format="{MemPercent}")` I get:
> 
> ```
> Traceback (most recent call last):
>   File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
>     self._context.run(self._callback, *self._args)
>   File "/usr/lib/python3.9/site-packages/libqtile/core/manager.py", line 913, in f
>     func(*args)
>   File "/usr/lib/python3.9/site-packages/libqtile/widget/base.py", line 457, in timer_setup
>     update_interval = self.tick()
>   File "/usr/lib/python3.9/site-packages/libqtile/widget/memory.py", line 59, in tick
>     self.update(self.poll())
>   File "/usr/lib/python3.9/site-packages/libqtile/widget/memory.py", line 76, in poll
>     return self.format.format(**val)
> KeyError: 'MemPercent'
> ```
> 
> Doesn't happen with the other Memory format options.

I had the same problem and I am running the version from github ( I don't remember when I downloaded and which branch/tag/commit I used but it was pretty recently ).

If you want to dirty fix it like I did and not depend on the package manager of your distro you need to use the `.percent` attribute from the object that `psutil.virtual_memory()` returns. You can actually change it since python is not a compiled language.

You need to go to the `widget/memory.py` file in your qtile installation folder. To find the path of the file just run the following:
python3 -c 'import libqtile;import os;print(os.path.join(os.path.dirname(libqtile.__file__), "widget", "memory.py"))'

Edit this file and add to the `val` `dict`
```
val["MemPercent"] = mem.percent
```
Like [this](https://github.com/qtile/qtile/blob/54a9b4f4d636ab0af54ca197ea80545a60b268f0/libqtile/widget/memory.py#L71)

**Comment by tych0** [comment]
On Wed, Dec 16, 2020 at 07:57:55AM -0800, Tilemachos Charalampous wrote:
> > When trying to call `widget.Memory(format="{MemPercent}")` I get:
> > 
> > ```
> > Traceback (most recent call last):
> >   File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
> >     self._context.run(self._callback, *self._args)
> >   File "/usr/lib/python3.9/site-packages/libqtile/core/manager.py", line 913, in f
> >     func(*args)
> >   File "/usr/lib/python3.9/site-packages/libqtile/widget/base.py", line 457, in timer_setup
> >     update_interval = self.tick()
> >   File "/usr/lib/python3.9/site-packages/libqtile/widget/memory.py", line 59, in tick
> >     self.update(self.poll())
> >   File "/usr/lib/python3.9/site-packages/libqtile/widget/memory.py", line 76, in poll
> >     return self.format.format(**val)
> > KeyError: 'MemPercent'
> > ```
> > 
> > Doesn't happen with the other Memory format options.
> 
> I had the same problem and I am running the version from github ( I don't remember when I downloaded and which branch/tag/commit I used but it was pretty recently ).
> 
> If you want to dirty fix it like I did and not depend on the package manager of your distro you need to use the `.percent` attribute from the object that `psutil.virtual_memory()` returns. You can actually change it since python is not a compiled language.
> 
> You need to go to the `widget/memory.py` file in your qtile installation folder. To find the path of the file just run the following:
> python3 -c 'import libqtile;import os;print(os.path.join(os.path.dirname(libqtile.__file__), "widget", "memory.py"))'
> 
> Edit this file and add to the `val` `dict`
> ```
> val["MemPercent"] = mem.percent
> ```
> Like [this](https://github.com/qtile/qtile/blob/54a9b4f4d636ab0af54ca197ea80545a60b268f0/libqtile/widget/memory.py#L71)

I would recommend against this. If you're having this problem with a
version of qtile that you believe has this patch, you should examine
your environment and see which version of qtile you are actually
running. It's likely that you have multiple versions installed, which
will only compound confusion.

**Comment by tchar** [comment]
> I would recommend against this. If you're having this problem with a version of qtile that you believe has this patch, you should examine your environment and see which version of qtile you are actually running. It's likely that you have multiple versions installed, which will only compound confusion.

Well I have been hacking into qtile for a week or so and as I was looking at the documentation for the Memory widget I tried the MemPercent but it was not working. I just went right into the code and saw that there was no MemPercent declared so I just added the line I described above without knowing that it actually exists in qtile master branch (I was thinking of making a pull request tbh).

It won't probably break anything since it is not working anyways but I get your point and you are right.

I am just providing an alternative in case @cjwyett does not want to download the master version because he is using a package manager with an older version that the master branch.
